onnxruntime/core/session/provider_bridge_ort.cc

```diff
@@ -0,0 +1,563 @@
+
+  Status Register(Provider_KernelCreateInfo&& create_info) override {
+    KernelCreateInfo info_real(std::move(static_cast<Provider_KernelDef_Impl*>(create_info.kernel_def.get())->p_),
+                               [kernel_create_func = create_info.kernel_create_func](const OpKernelInfo& info) -> OpKernel* {
+                                 Provider_OpKernelInfo_Impl provider_info(info);
+                                 return new OpKernel_Translator(provider_info, kernel_create_func(provider_info));
+                               });
+
+    return p_->Register(std::move(info_real));
+  }
+
+  std::shared_ptr<KernelRegistry> p_owned_;
+  KernelRegistry* p_{&*p_owned_};
+};
+
+struct Provider_IExecutionProvider_Router_Impl : Provider_IExecutionProvider_Router, IExecutionProvider {
+  Provider_IExecutionProvider_Router_Impl(Provider_IExecutionProvider* outer, const std::string& type) : IExecutionProvider(type), outer_(outer) {
+  }
+
+  virtual ~Provider_IExecutionProvider_Router_Impl() {}
+
+  std::shared_ptr<Provider_KernelRegistry> Provider_GetKernelRegistry() const override {
+    return std::make_shared<Provider_KernelRegistry_Impl>(GetKernelRegistry());
+  }
+
+  std::shared_ptr<KernelRegistry> GetKernelRegistry() const override {
+    return static_cast<Provider_KernelRegistry_Impl*>(&*outer_->Provider_GetKernelRegistry())->p_owned_;
+  }
+
+  std::vector<std::unique_ptr<Provider_ComputeCapability>> Provider_GetCapability(const onnxruntime::Provider_GraphViewer& graph,
+                                                                                  const std::vector<const Provider_KernelRegistry*>& kernel_registries) const override {
+    std::vector<const KernelRegistry*> kernel_registries_internal;
+    for (auto& v : kernel_registries)
+      kernel_registries_internal.emplace_back(static_cast<const Provider_KernelRegistry_Impl*>(v)->p_);
+
+    auto capabilities_internal = IExecutionProvider::GetCapability(static_cast<const Provider_GraphViewer_Impl*>(&graph)->v_, kernel_registries_internal);
+
+    std::vector<std::unique_ptr<Provider_ComputeCapability>> capabilities;
+    for (auto& v : capabilities_internal)
+      capabilities.emplace_back(onnxruntime::make_unique<Provider_ComputeCapability>(onnxruntime::make_unique<Provider_IndexedSubGraph_Impl>(std::move(v->sub_graph))));
+    return capabilities;
+  }
+
+  std::vector<std::unique_ptr<ComputeCapability>> GetCapability(const onnxruntime::GraphViewer& graph,
+                                                                const std::vector<const KernelRegistry*>& kernel_registries) const override {
+    std::vector<const Provider_KernelRegistry*> registries;
+    for (auto p : kernel_registries)
+      registries.push_back(new Provider_KernelRegistry_Impl(const_cast<KernelRegistry*>(p)));
+
+    auto provider_result = outer_->Provider_GetCapability(Provider_GraphViewer_Impl(graph), registries);
+    std::vector<std::unique_ptr<ComputeCapability>> result;
```

Comment:
result.reserve(provider_result.size()) and in other places as well where you're copying the output?

Reply:
Yeah, though in this case there's only one registry. I want to remove these layers entirely in the future as they're just pure overhead.

I did use reserve() in one place where it needs the addresses not to change later.